### PR TITLE
NameChoice (disambiguation): more refactoring

### PR DESCRIPTION
**Pull request by gasche**
This PR sits on top of of #9178. It contains 2 commits:
- "minor clarifications" has been written with @garrigue and @trefis;
   @Octachron, would you be willing to review it?
- "implementation comments" has been written with @garrigue and @Octachron;
   @trefis, would you be willing to review it?

### CODE REVIEW

**Comment by gasche** [comment]
Now that #9178 is merged, this PR is rebased on top of trunk and contains only the PR-specific changes.

File: typing/typecore.ml [hunk 1]
```diff
@@ -720,8 +721,9 @@ end) = struct
                                                 paths, false, expansion))
           end;
           lbl
-        with Not_found -> try
-          let lbl = lookup_from_type env tpath usage lid in
+        | exception Not_found ->
+        match lookup_from_type env tpath usage lid with
+        | lbl ->
```
Comment (trefis) [inline typing/typecore.ml]: I summon @Drup to fight this battle for me.

Comment (gasche) [reply typing/typecore.ml]: Taratata !

> "Justification-strengths should match change-importance." The strength of the justification required for a code-change decision is proportional to the importance of the change. A change with a large impact should be very strongly justified in discussions with reviewers, but for a nitpick-level change the author needs few justifications (possibly just personal preference) to apply or ignore a change suggestion.

This PR is joint work with @Octachron and @garrigue and both saw the code style in real time. OCtachron didn't say anything, and @garrigue approved.

File: typing/typecore.ml [hunk 2]
```diff
@@ -685,29 +692,33 @@ end) = struct
...
           let name = Datatype_kind.label_name kind in
           warn lid.loc
             (Warnings.Not_principal
                ("this type-based " ^ name ^ " disambiguation"))
         in
-        try
-          let lbl, use = disambiguate_by_type env tpath scope in
+        (* first look for a disambiguation solution
+           in the current lexical scope *)
+        match disambiguate_by_type env tpath scope with
```
Comment (gasche) [inline typing/typecore.ml]: this `match` needs a `begin` or a dedentation, my bad.

File: typing/typecore.ml [hunk 3]
```diff
@@ -780,6 +794,10 @@ module Label = NameChoice (struct
     | Record_unboxed true | Record_inlined _ | Record_extension _ -> false
 end)
 
+(* In record-construction expressions and patterns, we have many labels
+   at once; find a candidate type in the intersection of the candidates
+   of each label. In the [closed] expression case, this candidate must
+   contain exactly all the labels. *)
 let disambiguate_label_by_ids keep closed ids labels =
```
Comment (trefis) [inline typing/typecore.ml]: The explanation makes sense, and I'm fine with merging it as is.
But an explanation of `keep` as well as of the returned boolean would have been nice.

File: typing/typecore.ml (hunk 1 above)
Comment (trefis) [reply typing/typecore.ml]: Please.

> This does not mean that whitespace or comment changes can be made completely on whims, as different whitespace/comment changes have different impact and importance.

You're trying to sneak in a new way to indent things, and:
- everyone can already predict that you'll try to leverage it in other places later ("well if it was fine for that PR on typecore, why wouldn't it be nice for this other thing here...")
- it *has* an impact on readability.

You seem to have contradictory desires @gasche: on the one hand you want your refactor PRs to be merged quickly, on the other hand you insist on adding friction in every one of them.

:)

Comment (dbuenzli) [reply typing/typecore.ml]: There's an `.ocp-indent` file in the project isn't that supposed to be applied ? At least I did in [the past](https://github.com/ocaml/ocaml/pull/8834#discussion_r307978072) even if that was not to my liking.

File: typing/typecore.ml (hunk 3 above)
Comment (gasche) [reply typing/typecore.ml]: I just had a look. I think that it is covered by the following (existing) comment at the callsite:

```
       * if there is no known type reduce it incrementally, so that
         there is still at least one candidate (for error message)
```

The code proceeds by successive refinements/filterings of a list of "potential candidates". If I understand correctly, the idea is to stop reducing the set of possible candidates just before it becomes empty, and use the remaining candidates for suggestions in error messages. `keep` thus means something like "keep the result of the last filtering before the failure", and the boolean indicates whether it was a success. This could be usefully rewritten as a `(labels, labels) result`, with a comment that the Error case returns the last-nonempty list.

Comment (trefis) [reply typing/typecore.ml]: Thanks!
I think you're right that using a `(labels, labels) result` would be nicer.

Comment (gasche) [reply typing/typecore.ml]:
```ocaml
  match List.filter check_ids labels with
    | [] -> Error labels
    | labels ->
  match List.filter check_closed labels with
    | [] -> Error labels
    | labels ->
  Ok labels
```

**Comment by gasche** [comment]
I pushed a change suggested by @trefis and two further cleanups.

File: typing/typecore.ml (hunk 3 above)
Comment (lpw25) [reply typing/typecore.ml]: -1 to weird indentation style. If you're going to start suggesting this weird style in all you PRs please start a discussion on caml-devel about whether it is acceptable. Otherwise we're going to have this discussion on all of your PRs individually.

File: typing/typecore.ml [hunk 4]
```diff
@@ -680,15 +680,16 @@ end) = struct
                    usage lid env expected_type lbls =
     let scope = match scope with None -> lbls | Some l -> l in
     let warn_if_ambiguous lbl rest =
-      Printtyp.Conflicts.reset ();
```
Comment (trefis) [inline typing/typecore.ml]: Are you sure it's OK to push that call down? `ambiguous_types` does call `Printtyp` in some situations.
I guess @Octachron should have a look at this, as owner of the "conflicts" infrastructure.

Comment (Octachron) [reply typing/typecore.ml]: The function `Printtyp.Conflicts.reset ()` should indeed be called before `ambiguous_types` to start the conflict recording in `ambiguous_types` from a certainly clean state (or not at all under the supplementary assumption that all other users of conflicts cleaned the state after use) . Calling `Printtyp.Conflicts.reset` between `ambiguous_types` and `print_conflicts` erases any conflict explanations before they are printed.
I will send a test that catches this difference in behavior.

File: typing/typecore.ml [hunk 5]
```diff
@@ -670,92 +674,101 @@ end) = struct
         in
         List.find check_type lbls
 
+  (* scope: the labels equal to lid in the current lexical environment *)
+  (* lbls: possibly a subset of scope, see [disambiguate_label_by_ids] *)
   let disambiguate ?(warn=Location.prerr_warning) ?scope
-                   usage lid env opath lbls =
+                   usage lid env expected_type lbls =
```
Comment (garrigue) [inline typing/typecore.ml:726]: Could we replace the above comment by a real specification of this function:
```
(** [disambiguate] selects a concrete description for [lid] using some
    contextual information: an optional [expected_type], and a list of candidates [lbls].
    If [expected_type] is [None], it just returns the head of [lbls].
    Otherwise, it extracts the description directly from the corresponding type definition,
    except for extension types, where it has to choose it inside [lbls].
    This code is careful to warn in a number of situations:
    * if a non principal type was used for disambiguation [Not_principal]
    * if [lbls] contained several valid candidates [Ambiguous_name]
    * if the name was chosen out of [scope] [Name_out_of_scope] 
    Here [scope] contains the labels equal to [lid] in the current environment,
    and [lbls] is a (potentially strict) subset of [scope], see [disambiguate_label_by_ids] *)

File: typing/typecore.ml (hunk 4 above)
Comment (gasche) [reply typing/typecore.ml]: This is what I suspected, but I couldn't find an issue with my own manual inspection of the code, so I made the change in a commit marked for "separate review". I will get rid of the commit now.

File: typing/typecore.ml (hunk 5 above)
Comment (gasche) [reply typing/typecore.ml:726]: This looks nice overall but I have several questions on this proposed documentation. Could it maybe go in its own PR so that we can discuss it more easily? (That PR could be on trunk, or on top of this one, as you prefer.)

Comment (gasche) [reply typing/typecore.ml:726]: In fact I can also integrate it to my PR and ask questions on the diff directly. Will do.

File: typing/typecore.ml [hunk 6]
```diff
@@ -670,92 +674,109 @@ end) = struct
         in
         List.find check_type lbls
 
+  (** [disambiguate] selects a concrete description for [lid] using some
+      contextual information: an optional [expected_type], and a list of candidates [lbls].
```
Comment (gasche) [inline typing/typecore.ml]: @garrigue: in fact `lbls` is not a list but in some `result` type. Can you explain why?

File: typing/typecore.ml [hunk 7]
```diff
@@ -670,92 +674,109 @@ end) = struct
...
+      If [expected_type] is [None], it just returns the head of [lbls].
+      Otherwise, it extracts the description directly from the corresponding type definition,
+      except for extension types, where it has to choose it inside [lbls].
```
Comment (gasche) [inline typing/typecore.ml]: I don't see any special logic to handle extension types here, where does that come from? (Also, can we explain why?)

File: typing/typecore.ml [hunk 8]
```diff
@@ -670,92 +674,109 @@ end) = struct
...
+      This code is careful to warn in a number of situations:
+      * if a non principal type was used for disambiguation [Not_principal]
+      * if [lbls] contained several valid candidates [Ambiguous_name]
```
Comment (gasche) [inline typing/typecore.ml]: Why would that matter if type-based selection is used?

File: typing/typecore.ml (hunk 5 above)
Comment (garrigue) [reply typing/typecore.ml:726]: I would prefer to do it together, be cause the two are deeply related.

File: typing/typecore.ml [hunk 9]
```diff
@@ -670,92 +674,109 @@ end) = struct
...
+      * if the name was chosen out of [scope] [Name_out_of_scope]
+      Here [scope] contains the labels equal to [lid] in the current environment,
+      and [lbls] is a (potentially strict) subset of [scope], see [disambiguate_label_by_ids] *)
```
Comment (gasche) [inline typing/typecore.ml]: I think that these two remarks could be helpful to have *above* the rest of the specification, otherwise it's hard to follow the explanation if we assume that `lbls` is just an arbitrary list of labels (where are they coming from?).

Comment (gasche) [reply typing/typecore.ml]: Note: reading this API suggests that the code could be clearer if instead of taking a `scope, lbls` pair whose relation requires explanations, the code of `disambiguate` was parameterized over a filtering function `labels -> (labels, labels) result` that it would call itself on `scope`.

**Review by gasche — COMMENTED** [review]
Questions on @garrigue's proposal for a documentation comment of `disambiguate`.

File: typing/typecore.ml (hunk 6 above)
Comment (garrigue) [reply typing/typecore.ml]: Because, when the list is empty, we may have an error message, which we may want to use later.

File: typing/typecore.ml (hunk 7 above)
Comment (garrigue) [reply typing/typecore.ml]: Actually, the code first tries to look up the answer among lbls, so no specific code path is needed for that. The point I was making here is that, for all other types, the answer will be the one found through the type definition, and the change in order is only for warning purposes.

File: typing/typecore.ml (hunk 8 above)
Comment (garrigue) [reply typing/typecore.ml]: This matters because, when the type is not known principally, resolution without the type could have given a different result.

File: typing/typecore.ml (hunk 9 above)
Comment (garrigue) [reply typing/typecore.ml]: `scope` only appears on the previous line, and I'm not sure this information is needed to understand anything else.
The semantics of `disambiguate` does not require this inclusion, but if it is not true the fact we first look up among `lbls` matters (i.e. we could have a solution in `lbls` that would cause a scoping warning if reached through the type).

We can remove the explanation relating `lbls` and `scope`, and rewrite the explanation above to follow strictly the code:
```
Otherwise we try to find a description in `lbls` with a matching type,
or, if this fails, extract a description from the type definition itself.
```
Then we don't have to refer to extension types at all.

By the way, I forgot the last warning:
```
* if the selected name was not the same as before disambiguation was introduced [Disambiguated_name]
```

Comment (trefis) [reply typing/typecore.ml]: > We can remove the explanation relating lbls and scope, and rewrite the explanation above to follow strictly the code:
> ```
> Otherwise we try to find a description in `lbls` with a matching type,
> or, if this fails, extract a description from the type definition itself.
> ```
> Then we don't have to refer to extension types at all.

That would be a mistake: if you do that, you're not explaining anything, you're just paraphrasing the code. In this instance I think that the behavior was already obvious from the code (but others might disagree, in which case we can try and clarify *the code*), but I didn't know why that behavior was chosen.
A good *explanation* (mentioning extensible types) would be useful.

File: typing/typecore.ml [hunk 10]
```diff
@@ -670,92 +674,109 @@ end) = struct
...
+      warn lid.loc
+        (Warnings.Name_out_of_scope (path_s, [Longident.last lid.txt], false))
+    in
+    let warn_disambiguated_name lbl scope =
+      match scope with
+      | Ok ((lab1,_) :: _) when lab1 == lbl -> ()
+      | _ ->
+          Location.prerr_warning lid.loc
+            (Warnings.Disambiguated_name (get_name lbl))
+    in
```
Comment (trefis) [inline typing/typecore.ml]: FTR: I refrained from saying so earlier, but I'm curious to see what the code would look like if these helpers were lifted out. Right now they are introducing 30 lines of noise between the code and its explanation.

Comment (gasche) [reply typing/typecore.ml]: I pushed a commit that does it. It's not very nice because of the "warn" parameter -- it feels like code that was lambda-lifted by an automated tool rather than written by a human. I'm not convinced this is the best move, but let's wait to see how we like it.

File: typing/typecore.ml (hunk 6 above)
Comment (gasche) [reply typing/typecore.ml]: If I understand the code correctly, the error comes from some label-filtering logic that was done by the caller, and we could get rid of this complication by passing the filter to the function instead.

**Comment by gasche** [comment]
The discussion keeps suggesting small refactorings to do, but I am worried that we will end up with a PR with too many commits that result in a large diff that people don't want to review from scratch. 

Could we either:
- have people approve current commits, so that they can focus on reviewing only the new commits if we add more, or
- work on converging on the current PR, and delay further transformations to another PR?

File: typing/typecore.ml (hunk 9 above)
Comment (gasche) [reply typing/typecore.ml]: I agree with @trefis' point that the comment is closer to a description of the implementation (than to a specification). For me the parts that really helps reading the code is the explanation of what `lbls` and `scope` are (and maybe we could avoid need with clearer names, eg. `candidates_in_scope` and `candidates_selection`).

The part on the warnings is somewhat repetitive of the warnings themselves, I will get rid of it.

I'm interested in feedback on the other parts. (Personally I think that having some description is good, it could be shorter, but there is also some confusion, for example calling `lbls` a list when their type is not a list.)

File: typing/typecore.ml (hunk 7 above)
Comment (gasche) [reply typing/typecore.ml]: I still don't understand what happens for extension types and why that needs to be precised.

File: typing/typecore.ml (hunk 10 above)
Comment (garrigue) [reply typing/typecore.ml]: For once, I agree with @gasche: this code is about the warnings, which are only meaningful in that context, so moving them out does not improve things. Maybe we would be better off with a `where` syntax, but it was dropped in ocaml.

File: typing/typecore.ml (hunk 6 above)
Comment (garrigue) [reply typing/typecore.ml]: I'm afraid it would just increase the complication, as the caller would have to understand how this function is going to be called. But do we agree than any change in the API would have to be another PR anyway?

Comment (gasche) [reply typing/typecore.ml]: I asked about how to proceed with further changes [below](https://github.com/ocaml/ocaml/pull/9182#issuecomment-566671103). I agree with the idea of sending separate PRs, but I wished we would converge more quickly on each PR, because otherwise this could drag on indefinitely.

Concretely, would you be willing to approve the PR in its current state, even though we agree it's imperfect, so that follow-up PRs can start from `trunk` again? (I think we can iterate more on the documentation comment *and* I would like to experiment with the filtering logic, but those can go into separate PRs; the problem is that today they would conflict with the changes already in the PR.)

File: typing/typecore.ml (hunk 7 above)
Comment (garrigue) [reply typing/typecore.ml]: Just because for extension types there is no information in the type definitions, so the only possible approach is to collect the candidates in scope.
```
except for extension types, for which the information is not available,
so that one has to select it among [lbls] instead.
```

File: typing/typecore.ml (hunk 10 above)
Comment (gasche) [reply typing/typecore.ml]: (I think that @trefis' idea was also that sometimes, conciseness brings clarity, and the version with the warning code lifted out could result in a more brain-sized function that is easier to read and improve further.) I did think about using `let rec ... and ...` but I think it could be confusing.

File: typing/typecore.ml (hunk 9 above)
Comment (garrigue) [reply typing/typecore.ml]: OK, let's keep the old version.
And for the last two lines, I just copied what you had written :)
I'm fine with replacing `lbls` by `candidates`.
For `scope`, this is a bit more complicated. The only reason we keep `scope` independently of `lbls` is for cases where type-checking will eventually fail: a record construction with incompatible or missing labels, so that the label is eliminated from `lbls`, and rediscovered through the type. So I would not call its members `candidates_in_scope`, since they  were already discarded, and rather say `candidate_scope` if you want to be more explicit.

For the question about the use of a `return` type, I'm not sure it is essential to understand the code. I suppose it is clear that we are talking about the contents of the `return` type.
Maybe the person who introduced it can add a comment.

File: typing/typecore.ml (hunk 7 above)
Comment (gasche) [reply typing/typecore.ml]: What happens when an extension type is passed as expected type to `disambiguate`? The only control path I can see that matches your explanation is that it takes the `Some(tpath, tpath0, principal)` branch, for some reason `principal` is `false`, so we match on `lbls`, we take the first label suggested, it is in fact at the right path, it is not ambiguous, so no warning is shown.

Is that hypothesis correct?

Comment (gasche) [reply typing/typecore.ml]: (And also: if this reasoning holds, why would extension types satisfy `not principal` in the first place?)

File: typing/typecore.ml (hunk 9 above)
Comment (gasche) [reply typing/typecore.ml]: The result types were introduced by @lpw25 in c19e8b23506e38fc47c110138d50b57fe93f7d7e.

Comment (lpw25) [reply typing/typecore.ml]: Yes, it allows to keep all lookup logic in `Env`, previously it was spread all over the place. You call `lookup_all_foo` and get a `(foo list, lookup_error) result`. Either you have some `foo`s (ideally it would use a non-empty list type, but we don't have a convenient one of those lying around), or you have an error describing why we couldn't find any.

Comment (gasche) [reply typing/typecore.ml]: The point here, I think, is that the caller passes a `('a, error) result` instead of forcing the `'a` itself, because we don't know yet whether the value will in fact be needed (it is only needed if type-based disambiguation is not possible or is not principal), and we want to error only when necessary.

Comment (garrigue) [reply typing/typecore.ml]: Also, I don't think that removing the part on warnings is a good idea: a function name is not a specification. You might rather argue that my description was not precise enough.

Comment (lpw25) [reply typing/typecore.ml]: > and we want to error only when necessary.

That's right yes. Some of this complication is because the code mixes up the disambiguation code, which can work on long identifiers like `M.Foo`, with the code for type-based lookup, which only works on short identifiers like `Foo`. The long case potentially has interesting errors -- maybe it is `M` that doesn't exist rather than `Foo` -- but should always print the error if it exists. The short case has only a single error -- i.e. there is no `Foo` in scope -- but can ignore the error if it can find the constructor via the type.

File: typing/typecore.ml [hunk 11]
```diff
@@ -670,92 +674,114 @@ end) = struct
...
+
+  let warn_out_of_scope warn lid env tpath =
+    let path_s =
+      Printtyp.wrap_printing_env ~error:true env
+        (fun () -> Printtyp.string_of_path tpath) in
+    warn lid.loc
+      (Warnings.Name_out_of_scope (path_s, [Longident.last lid.txt], false))
+
+  (* warn if the selected name is not the last introduced in scope
+     -- in thes cases the resolution is different from pre-disambiguation OCaml
```
Comment (garrigue) [inline typing/typecore.ml]: `these`

**Comment by gasche** [comment]
I pushed a new PR, #9196, which implements my idea of turning the `lbls/scope` pair into a `scope / filter_candidates` pair. I do believe that it makes the code clearer.

Note: merging the current PR would make it sensibly easier to work on #9196 and also for you to review it. My personal assessment is that the remaining questions in the current discussion are small (or a matter of personal preference, should the warning functions be in or out?), compared to the improvement in the PR itself, so that we could and should stop here for this PR and iterate in further PRs. If you agree with this assessment, please approve the PR.

**Comment by trefis** [comment]
I'm willing to approve provided you fix the typo @garrigue just spotted. (And that you do not use the indentation style that is generating pushback? :>)

**Comment by trefis** [comment]
To make my last message clearer: I agree with your assessment that this PR is already a noticeable improvement.

File: typing/typecore.ml (hunk 7 above)
Comment (Octachron) [reply typing/typecore.ml]: I find the location of the comment about extension types a bit misleading: the `disambiguate` function has no special logic for them itself. It is the `Constructor` module that does not return any candidates for extension in its `lookup_all_from_type` implementations, and similarly it is `Env.lookup_all_constructors` that filters duplicate extension constructions from its return.

Comment (gasche) [reply typing/typecore.ml]: I improve on this in #9196 with a better comment and better placement.

**Review by garrigue — APPROVED** [review]
I'm happy with the current state, including the indentation.
But please get a consensus or revert to standard indentation before merging.

File: typing/typecore.ml [hunk 12]
```diff
@@ -778,21 +806,33 @@ module Label = NameChoice (struct
...
-  if keep && labels' = [] then (false, labels) else
-  let labels'' = List.filter check_closed labels' in
-  if keep && labels'' = [] then (false, labels') else (true, labels'')
+  match List.filter check_ids labels with
+  | [] -> Error labels
+  | labels ->
+  match List.filter check_closed labels with
+  | [] -> Error labels
+  | labels ->
+  Ok labels
```
Comment (Octachron) [inline typing/typecore.ml:832]: Why are the two checks not fused together into `List.filter (fun x -> check_ids x && check_closed x)`?

Comment (gasche) [reply typing/typecore.ml:832]: This is not equivalent because of the error discipline, where we return the last non-empty subset. If the first check is non-empty but the second is empty, your proposal would return the whole list instead of the first filter.

Comment (Octachron) [reply typing/typecore.ml:832]: Hm, I missed the shadowing of labels, this is a point against your indentation style here: the binding of labels should not be more indented that the body of the cases.

**Comment by Octachron** [comment]
Merged. This PR has been blocked for too long and there are worse flight of fancies than this ephemeral indentation style.
